File: Partition Labels.cpp
class Solution {
public:
    vector<int> partitionLabels(string s) {
        unordered_map<char, int> lastindex;
        vector<int> answer;
         for(int i=s.length()-1; i>=0; i--){
             if(lastindex.find(s[i])==lastindex.end())
                 lastindex[s[i]] = i;
         }
        int left=0, right=0;
        for(int i=0; i<s.length(); i++){
            right = max(right, lastindex[s[i]]);
            if(right==i){
                answer.push_back(i-left+1);
                left = i+1;
            }
        }
        return answer;
    }
};

File: Partition Labels.py
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        lastindex = defaultdict(int)
        partition = []

        for index in range(len(s)-1, -1, -1):
            if s[index] not in lastindex:
                lastindex[s[index]] = index

        start = 0
        while start < len(s):
            end = lastindex[s[start]]
            index = start
            while index < end:
                if lastindex[s[index]] > lastindex[s[end]]:
                    end = lastindex[s[index]]
                index += 1
            partition.append(end-start+1)
            start = end+1

        return partition
